### DISCUSSION TITLE
Adding a variable within an action

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#20666] [Q&A General] [answered]
Hello, I'm trying to add a bunch of variables within an action, this is a code snippet of how i'm doing it within the action under `_current_task == "add_variable"` in `act()`
for some reason, when I assign a kernel for any of these variables i get an error that the variable does not exist
this is confirmed by `_problem->hasVariable(variable_name)` which I run right after adding it to the variable.

```
InputParameters params = _factory.getValidParams("MooseVariableBase");
params.set<bool>("use_dual") = true;
params.set<MooseEnum>("family") = "SCALAR";
params.set<MooseEnum>("order") = "FIRST";
auto fe_type = AddVariableAction::feType(params);
auto var_type = AddVariableAction::variableType(fe_type);
_problem->addVariable(var_type, variable_name, params);
```

any suggestions of what could be the problem?

Thanks!

**Comment by KhaledNabilSharafeldin** [comment, marked as answer]
if the "family" param is changed to LAGRANGE, it seems to work.
any idea on how to initialise a SCALAR var?

**Reply by vanwdani** [reply]
I follow this procedure that should work for any order or family.

 
```
    {
      auto params = _factory.getValidParams("MooseVariable");
      const bool second = _problem->mesh().hasSecondOrderElements();
      if (!second)
        _order = "FIRST";
      params.set<MooseEnum>("order") = _order;
      params.set<MooseEnum>("family") = _family;
      params.set<std::vector<std::string>>("control_tags") = {"Variables"};

      _problem->addVariable("MooseVariable", "variable_name", params);
    }
```

**Reply by KhaledNabilSharafeldin** [reply]
I gave this a shot, but didnt work either.

I tried setting the type manually to `"MooseScalarVariable"` _which_ `auto var_type = AddVariableAction::variableType(fe_type);` _should determine on its own_.

i tried `_problem->addScalarVariable(var_name, _order, _scaling_factor, empty_vectorSubdomainID_ptr)`, adding the last parameter just adds the extra "block" parameter.

last thing I tried is getting all the SubdomainIDs `_problem->mesh().meshSubdomains();` then getting the corresponding names in a vector `_problem->mesh().getSubdomainName(subdomainID)` and adding the parameter "block".... but again! to no avail, still doesn't work!

it seems to be a problem with scalar family, when i changed it to LAGRANGE , it did set up the variables correctly.

still not sure where the problem is.

appreciate the help though!

**Reply by vanwdani** [reply]
`addScalarVariable` is deprecated and the use of `addVariable(var_type, var_name, params)` is recommended.  The block should be a parameter of the type.  Have you tried without `params.set<bool>("use_dual") = true`?

**Reply by GiudGiud** [reply]
this should work. Does it not?

```  
    auto lm_params = _factory.getValidParams("MooseVariableScalar");
    lm_params.set<MooseEnum>("family") = "scalar";
    lm_params.set<MooseEnum>("order") = "first";

    _problem->addVariable("MooseVariableScalar", "lambda", lm_params);
```

**Reply by KhaledNabilSharafeldin** [reply]
@GiudGiud no it doesn't but if i change the family it does work just fine

error for scalar variable:
```
*** ERROR ***
Scalar variable 'RVEBCACTION_LM0_dof0' does not exist in this system
```

**Reply by GiudGiud** [reply]
So when you change the family, you change the family of both the variable you are adding and the family that is expected by the kernel?

**Reply by KhaledNabilSharafeldin** [reply]
> `addScalarVariable` is deprecated and the use of `addVariable(var_type, var_name, params)` is recommended. The block should be a parameter of the type. Have you tried without `params.set<bool>("use_dual") = true`?

yes I was just testing various things to see what works, and yes I tried without "use_dual", doesnt make any difference.
and yes i did use the param<type> which is std::vector<SubdomainName>, i think it'll give a compile error if do otherwise.

**Reply by KhaledNabilSharafeldin** [reply]
> So when you change the family, you change the family of both the variable you are adding and the family that is expected by the kernel?

No, the kernel is a scalar kernel, but what happens is that all these NonlinearVariables do not have a variable and I am being prompted to add at least one kernel for them. that's how I know that at least they're registered for LAGRANGE family.

for scalar, even though i'm not adding it to a kernel yet, it prompts me that the variable does not exist

**Reply by KhaledNabilSharafeldin** [reply]
ok, adding the scalar variable works fine, but not adding a scalar kernel to a scalar variable give off the error stated before instead of the case with normal variables saying that there should be at least one kernel for that variable.

Thanks again for the help @vanwdani @GiudGiud

**Reply by GiudGiud** [reply]
Ah ok. The kernel_coverage_check for scalar variables has only been implemented for finite volume I think. 

That is a confusing error message that you got. I ll create an issue. Thanks for reporting it
